### PR TITLE
Symmetric KLIEP loss extension

### PR DESCRIPTION
**Pull request by byolkim**
The proposed changes implement the symmetric KLIEP loss.

### CODE REVIEW

File: src/utils_SymKLIEPLoss.jl [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+function CDSymKLIEPLoss(Ψx::Matrix{Float64}, Ψy::Matrix{Float64})
+    (p = size(Ψx, 1)) == size(Ψy, 1) || throw(DimensionMismatch())
+
+    CDSymKLIEPLoss(vec(mean(Ψx, dims=2)), vec(mean(Ψy, dims=2)), Ψx, Ψy, zeros(size(Ψx, 2)), zeros(size(Ψy, 2)), p)
+end
+
+CoordinateDescent.numCoordinates(f::CDSymKLIEPLoss) = f.p
+
+function CoordinateDescent.initialize!(f::CDSymKLIEPLoss, x::SparseIterate)
+    mul!(f.rx, transpose(f.Ψx),-x)
```
Comment (mlakolar) [inline src/utils_SymKLIEPLoss.jl]: This needs to create a temporary array for -x.
Better to multiply transpose(f.Ψx) with x.
Then change the sign of f.rx.

File: src/utils_SymKLIEPLoss.jl [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        @inbounds ry[i] += h*Ψy[j, i]
+    end
+    h
+end
+
+
+function SymKLIEP_Hessian(θ, Ψx, Ψy)
+    m, ny = size(Ψy)
+
+    wx =-transpose(Ψx) * θ
```
Comment (mlakolar) [inline src/utils_SymKLIEPLoss.jl]: add a space between = and - sign to make it clear what is going on

**Review by mlakolar — CHANGES_REQUESTED** [review]

File: src/utils_SymKLIEPLoss.jl (hunk 2 above)
Comment (byolkim) [reply src/utils_SymKLIEPLoss.jl]: done!

File: src/utils_SymKLIEPLoss.jl [hunk 3]
```diff
@@ -22,7 +22,9 @@ end
 CoordinateDescent.numCoordinates(f::CDSymKLIEPLoss) = f.p
 
 function CoordinateDescent.initialize!(f::CDSymKLIEPLoss, x::SparseIterate)
-    mul!(f.rx, transpose(f.Ψx),-x)
+    mul!(f.rx, transpose(f.Ψx), x)
+    f.rx *= -1.
```
Comment (mlakolar) [inline src/utils_SymKLIEPLoss.jl]: This should be 
`f.rx .*= -1.`

Notice the difference

```
julia> function f(a) 
         a *= -1.
         a
       end
f (generic function with 1 method)

julia> function g(a)
         a .*= -1.
         a
       end
g (generic function with 1 method)

julia> a = randn(1000000);

julia> @time f(a);
  0.005483 seconds (1.56 k allocations: 7.720 MiB)

julia> @time f(a);
  0.003113 seconds (6 allocations: 7.630 MiB)

julia> @time f(a);
  0.019711 seconds (6 allocations: 7.630 MiB, 83.54% gc time)

julia> @time g(a);
  0.017159 seconds (20.62 k allocations: 1.010 MiB)

julia> @time g(a);
  0.001316 seconds (4 allocations: 160 bytes)

julia> @time g(a);
  0.001246 seconds (4 allocations: 160 bytes)
```

The first time the function is compiled, but on the second run it is faster and allocates less memory.

**Review by mlakolar — CHANGES_REQUESTED** [review]

**Review by mlakolar — APPROVED** [review]
Great!

File: src/utils_SymKLIEPLoss.jl (hunk 3 above)
Comment (byolkim) [reply src/utils_SymKLIEPLoss.jl]: Changes made. Thanks! I'd never have imagined that a . could make such a difference.

Comment (mlakolar) [reply src/utils_SymKLIEPLoss.jl]: This is a useful link to read
https://docs.julialang.org/en/v1/manual/performance-tips/index.html

**Pull request merged**